src/interpreter/post_execution.rs

```diff
@@ -7,46 +7,44 @@ impl<S> Interpreter<S>
 where
     S: InterpreterStorage,
 {
-    /// Set the appropriate change output values after execution has concluded.
-    pub(crate) fn update_change_amounts(
-        &mut self,
-        unused_gas_cost: Word,
-        revert: bool,
-    ) -> Result<(), InterpreterError> {
+    /// Finalize outputs post-execution.
+    /// Set the appropriate change output values.
+    /// Revert variable output amounts as needed
+    /// TODO: do we want to set contract outputs in memory?
+    pub(crate) fn finalize_outputs(&mut self, unused_gas_cost: Word, revert: bool) -> Result<(), InterpreterError> {
         let init_balances = Self::initial_free_balances(&self.tx)?;
 
         let mut update_outputs = match &self.tx {
             Transaction::Script { outputs, .. } => outputs.clone(),
             Transaction::Create { outputs, .. } => outputs.clone(),
         };
+
         // Update each output based on free balance
-        for output in update_outputs.iter_mut() {
+        for (idx, output) in update_outputs.iter_mut().enumerate() {
             if let Output::Change { asset_id, amount, .. } = output {
                 let refund = if *asset_id == AssetId::default() {
                     unused_gas_cost
                 } else {
                     0
                 };
 
-                if revert {
-                    *amount = init_balances[asset_id] + refund;
+                let final_amount = if revert {
+                    init_balances[asset_id] + refund
```

Comment:
Yellow flag here: we will panic/halt (not talking about well-formed `RuntimeError`, but process crash) the VM if `asset_id` is not in `init_balances`. There is no hard guarantee (except for the transaction validity) that it will be there since the initial balances are calculated from the inputs, and here we are processing the outputs.

Do we want that?

Reply:
Realized this doesn't validate for coin outputs. Will add safety checks for this.